drivers/sqliter-driver/src/commonMain/kotlin/com/squareup/sqldelight/sqliter/driver/SQLiterHelper.kt

```diff
@@ -0,0 +1,228 @@
+                   private val database: DatabaseConnection) :
+        SqlPreparedStatement, EnforceClosed by EnforceClosedImpl() {
+    internal val statementList = frozenLinkedList<Statement>(stableIterator = false)
+    private val queryLock = QuickLock()
+    private val binds = frozenHashMap<Int, (Statement) -> Unit>()
+
+    internal fun close(){
+        queryLock.withLock {
+            checkNotClosed()
+            trackClosed()
+            statementList.forEach {
+                it.finalizeStatement()
+            }
+        }
+    }
+    override fun bindBytes(index: Int, bytes: ByteArray?) {
+        bytes.freeze()
+        binds.put(index) { it.bindBlob(index, bytes) }
+    }
+
+    override fun bindDouble(index: Int, double: Double?) {
+        binds.put(index) { it.bindDouble(index, double) }
+    }
+
+    override fun bindLong(index: Int, long: Long?) {
+        binds.put(index) { it.bindLong(index, long) }
+    }
+
+    override fun bindString(index: Int, string: String?) {
+        binds.put(index) { it.bindString(index, string) }
+    }
+
+    private fun bindTo(statement: Statement) {
+        for (bind in binds.values.iterator()) {
+            bind(statement)
+        }
+    }
+
+    override fun execute() = throw UnsupportedOperationException()
+
+    private fun findStatement():Statement = queryLock.withLock {
+        checkNotClosed()
+            return if (statementList.size == 0) {
+                database.createStatement(sql)
+            } else {
+                statementList.get(0)
+            }
+        }
+
+    internal fun cacheStatement(statement: Statement){
+        queryLock.withLock {
+            checkNotClosed()
+            statementList.add(statement)
+        }
+    }
+
+    override fun executeQuery(): SqlCursor {
+        val stmt = findStatement()
+        bindTo(stmt)
+        return SQLiterCursor(stmt, this)
+    }
+}
+
+class SQLiterStatement(private val statement: Statement, private val type: SqlPreparedStatement.Type) :
+        SqlPreparedStatement {
+
+    override fun bindBytes(index: Int, bytes: ByteArray?) {
+        bytes.freeze()
+        statement.bindBlob(index, bytes)
+    }
+
+    override fun bindDouble(index: Int, double: Double?) {
+        statement.bindDouble(index, double)
+    }
+
+    override fun bindLong(index: Int, long: Long?) {
+        statement.bindLong(index, long)
+    }
+
+    override fun bindString(index: Int, string: String?) {
+        statement.bindString(index, string)
+    }
+
+    override fun execute() {
+        when (type) {
+            SELECT -> throw AssertionError()
+            else -> statement.execute()
+        }
+    }
+
+    override fun executeQuery(): SqlCursor = throw UnsupportedOperationException()
+}
+
+class SQLiterCursor(val statement: Statement, private val query:SQLiterQuery) : SqlCursor {
+    private val cursor = statement.query()
+
+    override fun close() {
+        cursor.close()
+        query.cacheStatement(statement)
```

Comment:
This feels somewhat spaghetti-ish, but cursor needs it's own sqlite statement, which can be reused.